apps/agreement/contracts/Agreement.sol

```diff
@@ -932,304 +847,262 @@ contract Agreement is IArbitrable, IForwarder, AragonApp {
+        }
     }
 
     /**
-    * @dev Tell whether an address can sign the agreement or not
-    * @param _signer Address being queried
-    * @return True if the given address can sign the agreement, false otherwise
+    * @dev Change the challenge collateral of a disputable app
+    * @param _disputable Disputable app
+    * @param _disputableInfo Disputable info instance to change its collateral requirements
+    * @param _collateralToken Address of the ERC20 token to be used for collateral
+    * @param _actionAmount Amount of collateral tokens that will be locked every time an action is submitted
+    * @param _challengeAmount Amount of collateral tokens that will be locked every time an action is challenged
+    * @param _challengeDuration Duration in seconds of the challenge, during this time window the submitter can answer the challenge
     */
-    function _canSign(address _signer) internal view returns (bool) {
-        TokenBalancePermission storage permission = signTokenBalancePermission;
-        ERC20 permissionToken = permission.token;
+    function _changeCollateralRequirement(
+        IDisputable _disputable,
+        DisputableInfo storage _disputableInfo,
+        ERC20 _collateralToken,
+        uint256 _actionAmount,
+        uint256 _challengeAmount,
+        uint64 _challengeDuration
+    )
+        internal
+    {
+        require(isContract(address(_collateralToken)), ERROR_TOKEN_NOT_CONTRACT);
+
+        uint256 id = _disputableInfo.collateralRequirements.length++;
+        _disputableInfo.collateralRequirements[id] = CollateralRequirement({
+            token: _collateralToken,
+            actionAmount: _actionAmount,
+            challengeAmount: _challengeAmount,
+            challengeDuration: _challengeDuration
+        });
 
-        return isContract(address(permissionToken))
-            ? permissionToken.balanceOf(_signer) >= permission.balance
-            : canPerform(_signer, SIGN_ROLE, arr(_signer));
+        emit CollateralRequirementChanged(_disputable, id);
     }
 
     /**
-    * @dev Tell whether an address can challenge actions or not
-    * @param _challenger Address being queried
-    * @return True if the given address can challenge actions, false otherwise
+    * @dev Tell whether an address can challenge an action for a disputable app or not
+    * @param _disputable Disputable being queried
+    * @param _challenger Address of the challenger willing to challenge an action
+    * @return True if the challenger can be challenge actions on the disputable app, false otherwise
     */
-    function _canChallenge(address _challenger) internal view returns (bool) {
-        TokenBalancePermission storage permission = challengeTokenBalancePermission;
-        ERC20 permissionToken = permission.token;
+    function _canPerformChallenge(IDisputable _disputable, address _challenger) internal view returns (bool) {
+        if (!hasInitialized()) {
```

Comment:
For context, here we are checking against the Disputable app instead of the Agreement app. Thus, I needed to provide a custom implementation and I basically following the same way this is implemented for self-roles validations in `AragonApp` (see [here](https://github.com/aragon/aragonOS/blob/next/contracts/apps/AragonApp.sol#L42-L58)).

However, regarding the initialization check it's a good question I don't have a strong opinion for now. Maybe @sohkai can chime in. Adding that modifier to all functions I think it will basically make the contract size hit the limit.